typescript/looker/rtl/extensionSdk.ts

```diff
@@ -42,15 +44,28 @@ export interface IHostConnection {
   ): Promise<any>
 }
 
-// TODO update this to init31() and init40() explicit calls
 export class LookerExtensionSDK {
   /**
    * Creates a [[LookerSDK]] object.
+   *
+   * Examples:
+   * LookerExtensionSDK.createClient(host) => constructs a Looker31SDK
+   *
+   * LookerExtensionSDK.createClient(host, Looker40SDK) => constructs a Looker40SDK
    */
-  static createClient (hostConnection: IHostConnection) {
-    const settings = DefaultSettings()
+  static createClient<T extends APIMethods>(hostConnection: IHostConnection,
+                      type?: new (authSession: IAuthSession) => T,
+                      settings?: IApiSettings): T {
+    settings = settings || DefaultSettings()
     const transport = new ExtensionTransport(settings, hostConnection)
     const session = new ExtensionSession(settings, transport)
-    return new Looker31SDK(session)
+    if (type) {
+      return new type(session)
+    }
+    else
+    {
+      // work around TS2322
+      return ((new Looker31SDK(session)) as any) as T
```

Comment:
Does the consumer of this constructor provide type checking for API 4.0 methods vs. API 3.1 methods with this construction, or does this function think it returns `Looker31SDK | Looker40SDK`?

IOW, does

```typescript
let sdk = LookerExtensionSDK.createClient(host, Looker40SDK)
```

provide edit-time type checking for API 4.0?

If so, this looks good to me. If not, that's why we're using `.init31()` and `.init40()` methods for the other SDKs.

Reply:
The return type is `T extends APIMethods`, and the type passed as a parameter leads to type inferencing on the return type.  There is no union type.  In the code snippet you have above, `sdk` will be exactly `Looker40SDK` type.

This is also future-proof; it'll work for any future LookerSDK types, as long as they extend APIMethods.

Reply:
Excellent. Need to adopt that for other SDKs that can support it. Definitely for the other TS factories at least.